osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs

```diff
@@ -55,28 +42,49 @@ private void load()
                     Size = new Vector2(18),
                     Shadow = false,
                 },
+                loading = new DimmedLoadingLayer(0.8f, 0.5f),
             });
 
+            favourited.ValueChanged += favourited => icon.Icon = favourited.NewValue ? FontAwesome.Solid.Heart : FontAwesome.Regular.Heart;
+
             BeatmapSet.BindValueChanged(setInfo =>
             {
-                if (setInfo.NewValue?.OnlineInfo?.HasFavourited == null)
+                Enabled.Value = BeatmapSet.Value?.OnlineBeatmapSetID > 0;
+                favourited.Value = setInfo.NewValue?.OnlineInfo?.HasFavourited ?? false;
+            }, true);
+
+            Action = () =>
+            {
+                if (loading.State.Value == Visibility.Visible)
                     return;
 
-                favourited.Value = setInfo.NewValue.OnlineInfo.HasFavourited;
-            });
+                // guaranteed by disabled state abvove.
```

Comment:
I feel horrible for even suggesting this, but
```suggestion
                // guaranteed by disabled state above.
```